Mpi support

pmenczel opened the pull request:
**Summary**
This change adds `mpi_pmap` to qutip.parallel. It is an analogue of `parallel_map` that can run parallel computations on multiple CPUs (e.g., multiple supercomputer nodes) at the same time by using the MPI (message passing interface) API. We make use of the mpi4py module and its MPIPoolExecutor class. That is, parallel_map and mpi_pmap are basically the same, except that the `multiprocessing.ProcessPoolExecutor` gets replaced by the `mpi4py.MPIPoolExecutor`, and we allow users to provide arbitrary keyword arguments to be passed on to the MPIPoolExecutor.

**Detailed list of changes**
* Added `mpi_pmap`. Extracted the common functionality of parallel_map, loky_pmap and mpi_pmap into a `_generic_pmap` function. The behavior of parallel_map and loky_pmap should be unchanged (except for a tiny change in loky_pmap regarding the order in which reduce_func is called).
* Added a section showcasing QuTiP's MPI capabilities in the tutorial notebook "Non-Markovian Monte Carlo Solver", see PR [[TODO]] in qutip-tutorials.
* Added `mpi_options` to the options of MultiTrajSolver and all its subclasses. This option only takes effect if the 'map' option is set to 'mpi', in which case the mpi_options are added to the map_kwargs.
* For MultiTrajSolver, all its subclasses and corresponding XXsolve functions, I double-checked that the descriptions of the options in the docstrings correspond to the `solver_options`. That is, I did not touch the solver options except for adding 'mpi_options', but I edited the docstrings in some places in order to reflect the actually available solver options.
* While creating the new section in the tutorial notebook, I noticed some bugs in MultiTrajResult#__add__ and fixed them.
* Added tests. Tests are skipped if the mpi4py module is not available (like with loky_pmap and the loky module).

**Todo**
Include the changelog in a file named: `doc/changes/<PR number>.<type>` 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#changelog-generation) for more information).

**Related issues or PRs**
Continuing work started in [2280](https://github.com/qutip/qutip/pull/2280)

Ericgig reviewed (commented):
Thank you.

Could you add `mpi4py` to the requirement ins setup.cfg?

You added the function to tests, but mpi4py is never installed for the automated tests.
Did you try adding it to one of the job in  `.github/workflow/tests.yml`? I am pretty sure it will run with only one process, but running the parallel function on one cpu is better than not testing it at all. If you tried and couldn't install mpi4py on github runner, please add a note about the issues somewhere.

pmenczel commented on qutip/solver/mcsolve.py:
```diff
@@ -78,18 +76,26 @@ def mcsolve(H, state, tlist, c_ops=(), e_ops=None, ntraj=500, *,
         - | num_cpus : int
           | Number of cpus to use when running in parallel. ``None`` detect the
             number of available cpus.
+        - | bitgenerator : {None, "MT19937", "PCG64", "PCG64DXSM", ...}
+            Which of numpy.random's bitgenerator to use. With `None`, your
+            numpy version's default is used.
```
Just to clarify - you would leave out these options from the `mcsolve` etc docstrings, but still include them in the `MCSolver.options` etc docstrings?

Ericgig replied:
Yes

pmenczel commented on qutip/solver/multitraj.py:
```diff
@@ -143,10 +144,13 @@ def _initialize_run(self, state, ntraj=1, args=None, e_ops=(),
         map_func = _get_map[self.options['map']]
-        map_kw = {
+        map_kw = {}
+        if map_func == mpi_pmap:
+            map_kw.update(self.options['mpi_options'])
+        map_kw.update({
             'timeout': timeout,
             'num_cpus': self.options['num_cpus'],
-        }
+        })
```
My idea was that `mpi_options` should be ignored if it is provided but a map other than `mpi_pmap` is selected. That might be useful if a user is switching back and forth between `mpi_pmap` and `parallel_map` or so.

Ericgig replied:
I still believe these options should not be filtered here. If we add another map function that need more option, we would have to add another exception here. Also if we add parallel capacities elsewhere we would have to filter options the same way which will result in code duplication and potential mistake...

How about changing `_get_map` to be a function that take the options and return both `map_func` and `map_kw`?

pmenczel commented on qutip/solver/result.py:
```diff
@@ -829,14 +829,16 @@ def __add__(self, other):
             raise ValueError("Shared `times` are is required to merge results")
         new = self.__class__(self._raw_ops, self.options,
                              solver=self.solver, stats=self.stats)
+        new.e_ops = self.e_ops
```
As far as I can tell, the `e_ops` field was previously never set, because it doesn't get initialized in the constructor and only in `_add_first_traj` ([here](https://github.com/pmenczel/qutip/blob/57cca54994d2b70c9c99761b4726d1676c430b9e/qutip/solver/result.py#L506)). (I noticed it because if I write `result_sum = result1 + result2`, then `result_sum.__repr__()` fails because there is no `e_ops` attribute.)

I don't know why `e_ops` is set in `_add_first_traj` instead of in the constructor, so I left that unchanged.

pmenczel commented:
Thank you for the review.

> Could you add `mpi4py` to the requirement ins setup.cfg?
>
> You added the function to tests, but mpi4py is never installed for the automated tests. Did you try adding it to one of the job in `.github/workflow/tests.yml`? I am pretty sure it will run with only one process, but running the parallel function on one cpu is better than not testing it at all. If you tried and couldn't install mpi4py on github runner, please add a note about the issues somewhere.

I have added it to setup.cfg, and now some tasks are failing.
* The documentation build fails because it can't install `mpi4py` (`mpi.h` is missing). I don't quite understand why that is... I can install `mpi4py` on my laptop without problems (it just uses regular parallel processes instead).
* The `mpi_pmap` tests are skipped. I'm sorry, but I don't understand what needs to be done with the `tests.yml`.
* The failing tests on ubuntu-latest, python 3.10 / 3.11 seem unrelated to my changes.

Ericgig commented:
In the documentation, qutip is build with `pip install qutip[full]`, while in test it is installed with `pip install qutip[$QUTIP_TARGET]`, with the target changing depending on the test. We don't have any test using extra yet...

The mpi4py team made a GitHub Actions workflow to run it in test: https://github.com/mpi4py/setup-mpi.

You can add that workflow and the `extra` to have every test run with mpi:
```
      - uses: actions/checkout@v3
      - uses: mpi4py/setup-mpi@v1
```

`QUTIP_TARGET="tests,graphics,semidefinite,ipython,extra" `

By I am not sure it would work for windows. Also if installing mpi is slow, we may not want it in all test.
But conditional workflow are tricky so lets see want if it just works first.

Yes, the test failures are related to last weekend scipy release, not this code.

pmenczel commented:
Hmmm. Now the documentation build works, and the macos test run works as well (except for unrelated test failures).

The python 3.9 test runs fails to build mpi4py. Here, the problem solution would be [this](https://stackoverflow.com/a/73399242) (I've had the same problem before installing mpi4py on the RIKEN HPC). Should I add that?

All other test runs fail, seemingly because the first call to `mpi_pmap` hangs indefinitely. That's... hard to debug. Why would it work only on macos?

Ericgig commented:
Parallel map function having different behaviour on different OS is not new, it's quite annoying.
I am quite surprised (and sad) that it Mac that is going fine for this one. Most cluster are using linux, so I would be fine to just run the test once on linux...

I understand that you are developing with a mpi enable cluster, not just your local machine. Do you use `mpiexec` when running the tests? Maybe not using it with pytest could have side effects.

Maybe mpi4py's [discussions board](https://github.com/mpi4py/mpi4py/discussions) could help? They also have a google group, but it seems full of spam...

I have access to a cluster and I will try it later this week. I will come back to you if I find something interesting. But I have little experience with mpi in python.

pmenczel commented:
I am not using mpiexec when running the tests. I have done some more experimenting on my local machine.

1. If I install `mpi4py` with pip, I get the same problem on my machine (the test hangs, I can't even CTRL+C). The steps I have taken in detail are:
    ```bash
    conda create -n qutip-test-tmp python numpy scipy pytest
    conda activate qutip-test-tmp
    sudo apt install mpich libmpich-dev
    pip install 'qutip[full] @ git+https://github.com/qutip/qutip.git@refs/pull/2296/head'
    cd ~/miniconda3/envs/qutip-test-tmp/lib/python3.12/site-packages/qutip/tests/
    pytest --verbose --pyargs qutip.tests.solver.test_parallel
    ```
    (The `apt install` call is taken from the `mpi4py/setup-mpi` and without it, the installation fails. I `apt remove`d the packages again afterwards.)

2. Then I tried installing with conda as follows.
    ```bash
    conda create -n qutip-test2-tmp python numpy scipy pytest mpi4py
    conda activate qutip-test2-tmp
    pip install 'qutip[full] @ git+https://github.com/qutip/qutip.git@refs/pull/2296/head'
    cd ~/miniconda3/envs/qutip-test2-tmp/lib/python3.12/site-packages/qutip/tests/
    pytest --verbose --pyargs qutip.tests.solver.test_parallel
    ```
    Now things get even weirder. Here, `test_map` and `test_map_accumulator` pass, but `test_map_pass_error` hangs (I can get out with CTRL+C).

3. In my regular development environment, all tests pass without problem. I have no idea what the difference is between my regular environment and the one created in number 2. I will continue trying things out when I have time.

In the end, I imagine the solution might be not to include `mpi4py` in the `extras` section, and only add it with conda in one or two test runs?

I recently had a good experience asking for help on the mpi4py discussion board, might do that again.

Ericgig commented:
I tried running the tests on the local cluster. (Linux, openmpi/4, python 3.10)
It hanged on `test_map_pass_error` with `mpi4py.MPI.Exception: MPI_ERR_SPAWN: could not spawn processes`.

But running the tests with `mpiexec -n 1 pytest ...` worked fine.

pmenczel commented:
It seems to work now (the failing test is unrelated). I will briefly summarize what I found out and changed.

Setting up and using an MPI environment is surprisingly difficult, and I don't think it is in the scope of qutip to try and make that easier, or guide users to do that. The philosophy here is, if the user can get the `mpi4py.futures.MPIPoolExecutor` to work, then they can also use `mpi_pmap` in qutip. However, I did show in the tutorial notebook explicitly as an example how it works on the RIKEN cluster, and I refer to the mpi4py users guide.

That being said, we do want to set up an automated test for this, and we do want to understand what went wrong in our initial attempts to set up those tests. The two main problems we encountered were (a) if mpi4py is installed with pip, all mpi_pmap calls hang; (b) if it is installed with conda, test_map_pass_error hangs sometimes.
The difference between pip and conda comes from the fact that mpi4py requires an MPI implementation. If we use mpi4py's GH workflow to install it with pip, it installs MPICH, and with conda, it installs OpenMPI by default.

I noticed that in both cases where mpi_pmap hangs, there *are* error messages, but they are captured by pytest and then never printed because the program hangs... Regarding problem (a), in short, it looks like an MPICH bug to me. I thus decided to use OpenMPI instead, and added `conda install "openmpi<5" mpi4py` to just one of our test runs (the python 3.12 one).

Regarding problem (b), the problem was that we requested more processes than the default number of slots available. With OpenMPI, the default number of slots available is the number of physical cpu cores (which seems to be 2 on the machine running our tests, leaving only one slot for a worker process). In the `parallel` module, if num_cpus is not provided, it defaults to `available_cpu_count()` (which is 4 on this machine, since there are 4 logical cpus).

We could try to find a better default value for num_cpus for mpi_pmap, but I feel like writing that well would require deeper MPI knowledge than I have. What I did instead is, I tried to make it clear in the documentation that the default value might be unsuitable for MPI applications, and I emit a warning if the default value is used. At the same time, I set `OMPI_MCA_rmaps_base_oversubscribe=true` in the test run, which allows OpenMPI to create more processes than the number of available slots.

Ericgig commented on qutip/tests/solver/test_parallel.py:
```diff
@@ -22,14 +22,17 @@ def _func2(x, a, b, c, d=0, e=0, f=0):
-        loky = pytest.importorskip("loky")
+        pytest.importorskip("loky")
+    if map is mpi_pmap:
+        pytest.importorskip("mpi4py")
```
Will `2` cpus check be reliable with mpi?

pmenczel replied:
I think so. With setting `OMPI_MCA_rmaps_base_oversubscribe=true`, there should be no limit to the number of processes.

The only potential issue I can see is that the name of that environment variable was changed in OpenMPI 5. However, currently mpi4py only supports OpenMPI up to 4. Therefore I pinned the OpenMPI version to "<5".

Ericgig reviewed (commented):
Looking good.

I still believe that we should includ mpi4py in options.extras_require of `setup.cfg`.
Maybe it's own `mpi` section?

Ericgig reviewed (approved):
Thank you.
It's great to support mpi.
